code/modules/station_goals/dna_vault.dm

```diff
@@ -0,0 +1,301 @@
+
+/obj/structure/filler/ex_act()
+	return							
+							
+/obj/machinery/dna_vault
+	name = "DNA Vault"
+	desc = "Break glass in case of apocalypse."
+	icon = 'icons/obj/machines/dna_vault.dmi'
+	icon_state = "vault"
+	density = 1
+	anchored = 1
+	idle_power_usage = 5000
+	pixel_x = -32
+	pixel_y = -64
+	luminosity = 1
+
+	//High defaults so it's not completed automatically if there's no station goal
+	var/animals_max = 100
+	var/plants_max = 100
+	var/dna_max = 100
+	var/list/animals = list()
+	var/list/plants = list()
+	var/list/dna = list()
+
+	var/completed = FALSE
+	var/list/power_lottery = list()
+
+	var/list/obj/structure/fillers = list()
+
+/obj/machinery/dna_vault/New()
+	//TODO: Replace this,bsa and gravgen with some big machinery datum
+	var/list/occupied = list()
+	for(var/direct in list(EAST,WEST,SOUTHEAST,SOUTHWEST))
+		occupied += get_step(src,direct)
+	occupied += locate(x+1,y-2,z)
+	occupied += locate(x-1,y-2,z)
+
+	for(var/T in occupied)
+		var/obj/structure/filler/F = new(T)
+		F.parent = src
+		fillers += F
+
+	if(ticker.mode)
+		for(var/datum/station_goal/dna_vault/G in ticker.mode.station_goals)
+			animals_max = G.animal_count
+			plants_max = G.plant_count
+			dna_max = G.human_count
+			break
+
+/obj/machinery/dna_vault/Destroy()
+	for(var/V in fillers)
+		var/obj/structure/filler/filler = V
+		filler.parent = null
+		qdel(filler)
+	. = ..()
+	
+/obj/machinery/dna_vault/attack_ghost(mob/user)
+	if(stat & (BROKEN|MAINT))
+		return
+	return ui_interact(user)
+	
+/obj/machinery/dna_vault/attack_hand(mob/user)
+	if(..())
+		return 1
+	ui_interact(user)
+
+/obj/machinery/dna_vault/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1)
+	ui = nanomanager.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		roll_powers(user)
+		ui = new(user, src, ui_key, "dna_vault.tmpl", name, 550, 400)
+		ui.open()
+
+/obj/machinery/dna_vault/proc/roll_powers(mob/user)
+	if(user in power_lottery)
+		return
+	var/list/L = list()
+	var/list/possible_powers = list(VAULT_SPACEIMMUNE,VAULT_XRAY,VAULT_TELEKINESIS,VAULT_PSYCHIC,VAULT_SPEED)
+	L += pick_n_take(possible_powers)
+	L += pick_n_take(possible_powers)
+	power_lottery[user] = L
+
+/obj/machinery/dna_vault/ui_data(mob/user, ui_key = "main", datum/topic_state/state = physical_state) //TODO Make it % bars maybe
+	var/list/data = list()
+	data["plants"] = plants.len
+	data["plants_max"] = plants_max
+	data["animals"] = animals.len
+	data["animals_max"] = animals_max
+	data["dna"] = dna.len
+	data["dna_max"] = dna_max
+	data["completed"] = completed
+	data["used"] = TRUE
+	data["choiceA"] = ""
+	data["choiceB"] = ""
+	if(user && completed)
+		var/list/L = power_lottery[user]
+		if(L && L.len)
+			data["used"] = FALSE
+			data["choiceA"] = L[1]
+			data["choiceB"] = L[2]
+	return data
+
+/obj/machinery/dna_vault/Topic(href, href_list)
+	if(..())
+		return 1
+
+	if(href_list["gene"])
+		upgrade(usr,href_list["choice"])
+		. = TRUE
```

Comment:
What prevents a sneaky bastard from feeding a bad `Topic` url here and accessing the fruits of the vault early?